File: server/src/services/cache.js
// server/src/services/cache.js
// 精确缓存：相同的 system + message 直接返回缓存，不调 API
import crypto from 'crypto'
import { config } from '../config/index.js'

class ExactCache {
  constructor() {
    this.store = new Map()           // key → { content, ts, tokens }
    this.stats = { hits: 0, misses: 0, savedTokens: 0 }
  }

  // 用 MD5 把 system+message 变成 key
  _key(systemPrompt, message) {
    return crypto
      .createHash('md5')
      .update(`${systemPrompt || ''}||${message}`)
      .digest('hex')
  }

  get(systemPrompt, message) {
    const k = this._key(systemPrompt, message)
    const entry = this.store.get(k)

    if (!entry) {
      this.stats.misses++
      return null
    }

    // 检查是否过期
    if (Date.now() - entry.ts > config.cache.ttl) {
      this.store.delete(k)
      this.stats.misses++
      return null
    }

    this.stats.hits++
    this.stats.savedTokens += entry.tokens || 0
    return entry
  }

  set(systemPrompt, message, { content, tokens = 0 }) {
    const k = this._key(systemPrompt, message)
    this.store.set(k, { content, tokens, ts: Date.now() })

    // 缓存超过 500 条时，清除最老的一批（简单 LRU）
    if (this.store.size > 500) {
      const oldestKeys = [...this.store.entries()]
        .sort((a, b) => a[1].ts - b[1].ts)
        .slice(0, 50)
        .map(([k]) => k)
      oldestKeys.forEach(k => this.store.delete(k))
    }
  }

  // 命中率
  get hitRate() {
    const total = this.stats.hits + this.stats.misses
    return total === 0 ? '0%' : `${(this.stats.hits / total * 100).toFixed(1)}%`
  }

  getStats() {
    return {
      size:        this.store.size,
      hits:        this.stats.hits,
      misses:      this.stats.misses,
      hitRate:     this.hitRate,
      savedTokens: this.stats.savedTokens,
    }
  }
}

// 全局单例
export const cache = new ExactCache()

File: server-py/app/services/cache.py
# server-py/app/services/cache.py
# 精确缓存：相同的 system + message 直接返回缓存，不调 API
import hashlib
import time

from app.config import config


class ExactCache:
    def __init__(self):
        self.store: dict[str, dict] = {}  # key → { content, ts, tokens }
        self.stats = {"hits": 0, "misses": 0, "savedTokens": 0}

    def _key(self, system_prompt: str | None, message: str) -> str:
        raw = f"{system_prompt or ''}||{message}"
        return hashlib.md5(raw.encode("utf-8")).hexdigest()

    def get(self, system_prompt: str | None, message: str):
        k = self._key(system_prompt, message)
        entry = self.store.get(k)

        if not entry:
            self.stats["misses"] += 1
            return None

        # 毫秒级 TTL（与 JS 版保持一致）
        if (time.time() * 1000) - entry["ts"] > config.cache.ttl:
            del self.store[k]
            self.stats["misses"] += 1
            return None

        self.stats["hits"] += 1
        self.stats["savedTokens"] += entry.get("tokens", 0)
        return entry

    def set(self, system_prompt: str | None, message: str, content: str, tokens: int = 0):
        k = self._key(system_prompt, message)
        self.store[k] = {"content": content, "tokens": tokens, "ts": time.time() * 1000}

        # 缓存超过 500 条时，清除最老的一批（简单 LRU）
        if len(self.store) > 500:
            oldest = sorted(self.store.items(), key=lambda kv: kv[1]["ts"])[:50]
            for key, _ in oldest:
                del self.store[key]

    @property
    def hit_rate(self) -> str:
        total = self.stats["hits"] + self.stats["misses"]
        return "0%" if total == 0 else f"{self.stats['hits'] / total * 100:.1f}%"

    def get_stats(self) -> dict:
        return {
            "size": len(self.store),
            "hits": self.stats["hits"],
            "misses": self.stats["misses"],
            "hitRate": self.hit_rate,
            "savedTokens": self.stats["savedTokens"],
        }


# 全局单例
cache = ExactCache()
